include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -2982,6 +2982,56 @@ namespace internal
 
 
 
+  template <int dim, int length, typename Number2, typename Number>
+  inline DEAL_II_ALWAYS_INLINE
```

Comment:
Does this need to be `DEAL_II_ALWAYS_INLINE`?

Reply:
What speaks against `DEAL_II_ALWAYS_INLINE` from your perspective? It definitely should be inlined, the goal of this PR is to improve this performance-critical part of the code.

Reply:
We typically don't force inlining (i.e., we let the compiler decide) unless the function is very small - I would expect using `DEAL_II_ALWAYS_INLINE` here to actually make things slower. It might be worth trying and measuring to see which is faster.

Reply:
I agree, we should take a look. It depends on how much function prologue/epilogue is necessary and how the compiler decides to realize the inlining. Compiler can make pretty bad decisions sometimes, because it is hard to know what is really an inner loop (thus worth inlining) and what gets called less often and should limit the code size instead (better not to inline). The current suggestion is based on past experience with compilers, so it might be different here. @bergbauer can you please collect the run time for your benchmark and a range of degrees with the two options? Either way, we should use something like https://github.com/dealii/dealii/blob/b8135fa6eb6e8b59cf9aab7b5e2ded518d2fdf61/include/deal.II/matrix_free/evaluation_kernels_hanging_nodes.h#L31-L35

Reply:
![Screenshot from 2023-03-27 11-12-56](https://user-images.githubusercontent.com/44996996/227897552-2609f912-dfb7-4001-8c66-c7fd098264e8.png)

I collected the throughput and instruction counts for the two relevant functions, you cannot see much of a difference between `inline` and `DEAL_II_ALWAYS_INLINE`. What do you think, should we go for plain `inline`?

Reply:
The difference seems to be in the noise. So let us go for the code without `DEAL_II_ALWAYS_INLINE` as suggested by @drwells, and only add in case we have a case with a problem. Thank you for measuring the performance.

Reply:
Thanks for making me aware of the difference between `inline` and `DEAL_II_ALWAYS_INLINE`. Done.